lib/JSON/Validator.pm

```diff
@@ -573,69 +573,85 @@ sub _validate {
 
 sub _validate_all_of {
   my ($self, $data, $path, $rules) = @_;
-  my $type = data_type $data, $rules;
-  my (@errors, @expected);
+  my (@errors);
 
   my $i = 0;
   for my $rule (@$rules) {
     next unless my @e = $self->_validate($_[1], $path, $rule);
-    my $schema_type = schema_type $rule;
-    push @expected, $schema_type if $schema_type;
-    push @errors, [$i, @e] if !$schema_type or $schema_type eq $type;
+    push @errors, [$i, @e];
   }
   continue {
     $i++;
   }
 
-  return E $path, [allOf => type => join('/', uniq @expected), $type]
-    if !@errors and @expected;
-  return prefix_errors allOf => @errors if @errors;
-  return;
+  return if not @errors;
+
+  return E $path,
+    [
+    allOf => type => join('/', uniq map $_->[1]->details->[0], @errors),
+    $errors[-1][1]->details->[2]
+    ]
+    if @errors > 1 and not grep $_->[1]->details->[1] ne 'type', @errors;
+
+  return prefix_errors allOf => @errors;
 }
 
 sub _validate_any_of {
-  my ($self, $data, $path, $rules) = @_;
-  my $type = data_type $data, $rules;
-  my (@e, @errors, @expected);
+  my ($self, $data, $path, $rules, $prefix) = @_;
+  $prefix //= 1;
```

Comment:
I don't like this `$prefix` variable. It's a bit worse than my `$expected` variable in `_validate_type_number()` 😥 Can we make prefix_errors() smarter instead?

Reply:
I don't either, but it was late when I got to this point and I kind of gave up. :(  The key here is indicating to _validate_any_of whether it should actually be prefixing a path to the errors it got, or not -- when called via `type => [...]` there isn't any literal anyOf rule from the user, so we should just report the error(s) as-is without a prefix.